1257 allow selection of units in select metadata menu

juzen2003 opened the pull request:
- Fixes #1257 (implementation in select metadata menu)
- Were any Django, import pipeline, or table_schema files modified? N
  - If YES:
    - Database used: opus3_test_220906
    - All Django tests pass: Y
    - FLAKE8 run on modified code: Y
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA
- Was the documentation reviewed for necessary changes or additions? Y
  - About
  - Getting Started
  - FAQ
  - API Guide
  - Tooltips

Description of changes:
- Add units with dropdown in selected metadata fields when:
    - The select metadata menu is open
    - A new field is added
        - Store current displayed unit and available units in the data attributes for each menu item so that later on when a new field is added, we can use these info to attached the unit dropdown to that specific field
- Style the dropdown
    - Set the max height
    - Enable scrolling
- Add the dropdown behaviors
    - Scroll on dropdown won't scroll the selected metadata field
    - Click dropdown item will:
        - Close the dropdown
        - Change the displayed unit 
        - Update the slug in opus.prefs.cols
            - Add ":" to specify unit, default unit will be used if it's not specified.
            - Check ":" when parsing slugs from opus.prefs.cols.  
        - Avoid page refreshing 
- Click "Save Changes" will update the metadata values based on selected units

juzen2003 committed:
- Add units with dropdown in selected metadata fields when the select
  metadata menu is open.
- Store default unit and available units in the data attributes for each
  menu item so that later on when a new field is added, we can use these
  info to attached the unit dropdown to that specific field.
- Style the dropdown to enable scrolling.
- Avoid scrolling  the selected metadata field list when scrolling on
  the unit dropdown.

juzen2003 committed:
Remove commented out code in ui/views.py.

juzen2003 committed:
Add units with dropdown in selected metadata fields when a new field is
added in selected metadata menu.

juzen2003 committed:
- Avoid URL change when clicking an dropdown item
- Change the displayed unit for a selected metadata field when selecting
  a new unit in the dropdown menu.

juzen2003 committed:
Update opus.prefs.cols when an item in unit dropdown is selected.

juzen2003 committed:
Check ":" when parsing slugs from opus.prefs.cols.
- In views.py (line 208-218), displayed data with the desired unit in
  metadata if the desired unit is provided.
- In utils.js (line 172-177), create a function getSlugWithoutUnit to
  return the slug without a unit. This is used when we are parsing slugs
  from opus.prefs.cols.
- In selectMetadata.js/browse.js, call getSlugWithoutUnit in to handle ":"
  in the slugs from opus.prefs.cols (multiple places)
- In menu.html & select_metadata.html, rename data attribute from
  "defaultunit" to "dispunit"

juzen2003 committed:
Remove unused import function "get_unit_display_name" in ui/views.py

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 1257_allow_selection_of_units_in_UI

rfrenchseti commented on lib/opus_support.py:
```diff
@@ -1798,6 +1798,20 @@ def display_unit_ever(unit_id):
+        default_unit_disp = get_unit_display_name(form_type_unit_id, default_unit)
+    else:
+        default_unit_disp = None
+    return [default_unit_disp, available_units]
```
No need to return a list...you can return a tuple since it's more efficient. Just remove the [].

rfrenchseti commented on opus/application/apps/paraminfo/models.py:
```diff
@@ -118,9 +117,17 @@ def get_units(self, override_unit=None):
+        display_name = self.get_default_unit(override_unit)
+        if display_name:
+            return ('(' + display_name + ')')
+        else:
```
Remove else and just always return display_name

rfrenchseti commented:
Also need to add a brief mention of this functionality in the Getting Started Guide. Add a new sentence after this:

"Which metadata fields are displayed in the table view and slideshow view can be changed by clicking on Select Metadata at the top. Fields may be added or removed by clicking on them on the left or removed by clicking on the on the right, and may be reordered by dragging."

First of all fix the existing typo: "clicking on THEM on the right"

Then add: "If available, the units a metadata field is displayed in can be changed by clicking on the current units name shown in parentheses."

The next sentence begins "The fields selected here also affect". Change to "The fields and units selected here also affect"

juzen2003 committed:
- Remove ":unit" in cols when the default unit is selected in unit dropdown
  menu.
- Clean up code style and comments based on github feedback.

juzen2003 committed:
Update the getting started guide based on github feedback (for units
dropdown menu in select metadata menu)

juzen2003 committed:
Rename function get_disp_and_available_units to get_disp_default_and_avail_units

rfrenchseti reviewed (changes requested):
If you add a field that uses units purely for format conversion, like Cassini Spacecraft Clock Start Count, it looks fine. But if you close Select Metadata and then open it again, it says (NULL) in the units position.

juzen2003 committed:
Update get_disp_default_and_avail_units, for units purely for format
conversion, we would not display it. (line 1805-1809, opus_support.py)

rfrenchseti reviewed (changes requested):
When you drag-move an existing column in Select Metadata, it loses the information about which units have been selected (only visible after saving the state).

juzen2003 committed:
Update metadataDragged to add ":unit" for slugs in cols with units selected.
This fixed the issue that selected units are ignored after selected metadata
is re-ordered. (line 375-390, selectMetadata.js)

dstopp commented on opus/application/apps/help/templates/help/gettingstarted.html:
```diff
@@ -220,13 +220,15 @@ <h3>Browsing</h3>
     on the <i class="far fa-trash-alt"></i> on the right, and may be reordered
-    by dragging. Fields can be added or removed in the table view by
-    hovering over the column header; click on the <i class="far fa-trash-alt"></i>
+    by dragging. If available, the units a metadata field is displayed in can be
```
I think this may be better worded like this:

When available, the display units for the metadata field can be changed by clicking on the current units name shown in the parentheses.

rfrenchseti replied:
I think I would get rid of "the" and say: When available, the display units for the metadata field can be changed by clicking on the current units name shown in parentheses.

dstopp commented on lib/opus_support.py:
```diff
@@ -1798,6 +1798,23 @@ def display_unit_ever(unit_id):
     return display_search_unit(unit_id) or display_result_unit(unit_id)
 
+def get_disp_default_and_avail_units(param_form_type):
+    "Return display, default, and available units for a given param form type."
```
please change 'param' to 'ParamInfo' for clarity in comment

dstopp commented on opus/application/apps/ui/templates/ui/menu.html:
```diff
@@ -46,7 +47,11 @@
-                                                      " data-slug="{{ p.slug }}" class="op-search-param" href="#">
+                                                      "
+                                                      data-dispunit="{% if p.disp_unit %}{{ p.disp_unit }}{% endif %}"
+                                                      data-defaultpunit="{% if p.default_unit %}{{ p.default_unit }}{% endif %}"
```
I believe data-defaultpunit is a typo; seems it should likely be data-defaultunit.  Suggestion is to change these like this:
data-dispunit  :     data-dispUnits
data-defaultunit:  data-defaultUnits
data-availunits:     data-availUnits

dstopp commented on opus/application/static_media/js/selectMetadata.js:
```diff
@@ -290,14 +316,32 @@ var o_selectMetadata = {
+            <div class="dropdown-menu op-scrollable-menu" aria-labelledby="dropdownMenuLink">`;
+            let units = $(menuSelector).data("availunits");
+            for (let unit in units) {
+                unitDropdown += `<a class="dropdown-item" data-defaultunit="${defaultUnit}" data-value="${unit}" data-dispval="${units[unit]}" data-slug="${slug}" href="#">${units[unit]}</a>`;
```
Adding data-dispval to the possible list of rethinking,   where data-dispValue may be more consistent <at least w/the other data things that are being used for this feature - ie., if we change those, then we should change this as well>

juzen2003 committed:
Modify & update variable names based on github feedback.

juzen2003 committed:
Remove the debug=false option in the browse view tooltipster.

dstopp reviewed (changes requested):
Pull down stays open when it should close - so far, i've found two ways:
- Open Select Metadata;
- Pull down one of the menus;
- Close the window w/out saving
- Reopen and the menu is still there
OR
- Open Select Metadata; 
-  pull down one of the units; delete that metadata w/the trash can.  
- Select discard changes
- Open Select metadata - 
- the pull down will still be open
- if you open a new pull down, it will overlay the one that came back and I couldn't find a way to close it other than refresh

juzen2003 committed:
- Close units dropdown when:
    - select metadata modal is closed
    - drag any selected metadata fields
    - click on trash icon of the selected metadata fields
- Restore the display unit on selected metadata fields when the changes
  are not saved and modal is closed

juzen2003 committed:
Merge remote-tracking branch 'origin/1257_allow_selection_of_units_in_UI' into 1257_allow_selection_of_units_in_UI

juzen2003 committed:
- Rename "data-dispUnit" to "data-dispunit", "data-defaultUnit" to
  "data-defaultunit", and "data-availUnits" to "data-availunits".
- Remove the console warning messages of trying to apply tooltipster on
  the same element for multiple times.

dstopp commented on opus/application/static_media/js/selectMetadata.js:
```diff
@@ -290,14 +330,32 @@ var o_selectMetadata = {
+            <div class="dropdown-menu op-scrollable-menu" aria-labelledby="dropdownMenuLink">`;
+            let units = $(menuSelector).data("availunits");
+            for (let unit in units) {
+                unitDropdown += `<a class="dropdown-item" data-defaultunit="${defaultUnit}" data-value="${unit}" data-dispValue="${units[unit]}" data-slug="${slug}" href="#">${units[unit]}</a>`;
```
I think you changed data-dispValue to data-dispvalue

dstopp commented on opus/application/apps/ui/templates/ui/select_metadata.html:
```diff
@@ -39,7 +39,20 @@
+                        <span>{{ param_info.body_qualified_label_results }}
+                        {% if param_info.units %}
+                            (<div class="op-units-dropdown dropdown d-inline">
+                                <a class="op-{{ param_info.slug }}-units-dropdown-toggle dropdown-toggle" href="#" role="button" id="dropdownMenuLink" data-toggle="dropdown" data-prevDispValue="{{ param_info.disp_unit }}" aria-haspopup="true" aria-expanded="false">
```
I think you changed data-prevDispValue to data-prevdispvalue

dstopp reviewed (changes requested):
I think the data-{item} things are not yet consistently all lower case or all camel case.  After some conversation, my understanding was to leave them as all lower case but I still see some discrepancies in the code.

juzen2003 committed:
Change "data-prevDispValue" to "data-prevdisvalue"

juzen2003 committed:
Change all camel case data-* attribuet to lowercase in select_metadata.html

juzen2003 committed:
Change "data-dispValue" to "data-dispvalue" at line 346 in selectMetadata.js
